// openrouter-sdk/src/index.ts
/**
 * OpenRouter SDK - Main exports
 */

// Core components
export { OpenRouter } from './core/open-router.js';
export { AIOrchestrator } from './core/ai-orchestrator.js';

// Interfaces
export * from './interfaces/index.js';

// Utilities
export { Logger } from './utils/logger.js';
export { MemoryCache } from './utils/memory-cache.js';
export { RateLimiter } from './utils/rate-limiter.js';
export { FunctionCalling } from './utils/function-calling.js';
export { VectorDB } from './utils/vector-db.js';
export { CrewAI } from './utils/crew-ai.js';
export { AgentMemory, MemoryType } from './utils/agent-memory.js';
export { WebSearch } from './utils/web-search.js';
export { StructuredOutput } from './utils/structured-output.js';
export { Reasoning } from './utils/reasoning.js';
export { ProviderRouting } from './utils/provider-routing.js';

// Providers
export { ClaudeProvider, ClaudeConfig } from './providers/claude.js';

// Services
export { GoogleSearch, GoogleSearchConfig, SearchResult } from './services/google-search.js';

// Examples
// Note: Enhanced crew and agent memory examples were removed

// Errors
export { OpenRouterError } from './errors/openrouter-error.js';
